### PR TITLE
Adds a simple payment component, photocopiers now cost 5 credits to make copies each.

### LINKED ISSUE #47853: Photocopiers only work for single copies
## Testmerges:
didn't check lol

## Reproduction:
put a paper inside a photocopier and try making it copy it more than once, only one copy will come out, if you hit copy again you get another copy, making it work but only one at a time, when you used to be able to make it make many copies without having to hit copy every time.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
I beg to hear your ass photo bikeshedding.

## About The Pull Request

This adds a new component called payment.  It's useful for objects that charge the user for a uniform, standard fee for preforming an action. Should be simple enough that it could be expanded upon and save some space on the 100 other things have unique behavior for how and why to charge someone for an action/object. In it's current iteration, it's best for simple static price fees.

As for the gameplay differences here, this alters photocopiers so that every new copy costs a flat 5 credits to the crewmember, and pays out to the civilian budget.

Update: In addition to the above, I have also fixed #47853.

## Why It's Good For The Game

I am 100% at fault of this, in that there's about 12 different things in game which charge or pay players, and do so in very similar ways with slight variations. This is a pretty basic framework for an easy payment, so basic costs and tolls can be automated to better integrate things objects with the cash economy.

## Changelog
:cl:
tweak: The photocopier now has a 5 credit cost to use.
code: There is now a uniform component for simple payments.
fix: You can now make copies of paper with the photocopier again.
/:cl:

### CODE REVIEW

**Comment by TemporalOroboros** [comment]
This is good, but I have to ask; why are you creating a signal _specifically_ for the printer? Shouldn't there just be the one signal for attempting to charge or am I missing something.

Also, it would be nice for you to add some logging if that isn't taken care of already by other code.

File: code/datums/components/payment.dm [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+/datum/component/payment/Initialize(_cost, _target, _style)
+	if(!_target)
+		target = SSeconomy.get_dep_account(ACCOUNT_CIV)
+	target = _target
+	cost = _cost
+	transaction_style = _style
+	RegisterSignal(parent, list(COMSIG_OBJ_ATTEMPT_CHARGE_PRINTER, COMSIG_OBJ_ATTEMPT_CHARGE), .proc/attempt_charge)
+
+/datum/component/payment/proc/attempt_charge(datum/source, atom/movable/AM)
```
Comment (bobbahbrown) [inline code/datums/components/payment.dm:27]: The return type of a signal should be bitflags.

[You can see that boolean return values are unacceptable in this post here by Ninja.](https://tgstation13.org/phpBB/viewtopic.php?f=5&t=22674&p=524299#p494359)

File: code/datums/components/payment.dm [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  * Objects that can drain the owner's money linearly.
+  * What this is not useful for:
+  * Things where the seller may want to fluxuate the price of the object.
+  * Improving standardizing every form of payment handing, as some custom handling is specific to that object.
+  **/
+/datum/component/payment
+	///Standardized of operation.
+	var/cost = 10
+	///Flavor style for handling cash (Friendly? Hostile? etc.)
+	var/transaction_style = "Clinical"
```
Comment (bobbahbrown) [inline code/datums/components/payment.dm:15]: This should be a define

File: code/datums/components/payment.dm [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		return FALSE
+	if(!(card.registered_account.has_money(cost)))
+		switch(transaction_style)
+			if("Friendly")
+				to_chat(user, "<span class='warning'>I'm so sorry... You don't seem to have enough money.</span>")
+			if("Angry")
+				to_chat(user, "<span class='warning'>YOU MORON. YOU ABSOLUTE BAFOON. YOU INSUFFERABLE TOOL. YOU ARE POOR.</span>")
+			if("Clinical")
+				to_chat(user, "<span class='warning'>ID Card lacks funds. Aborting.</span>")
+		return FALSE
```
Comment (bobbahbrown) [inline code/datums/components/payment.dm:58]: Similar to above, these should be defined constants (BEHAVIOUR_FRIENDLY, etc etc).

File: code/modules/paperwork/photocopier.dm [hunk 4]
```diff
@@ -29,6 +29,10 @@
 	var/mob/living/ass //i can't believe i didn't write a stupid-ass comment about this var when i first coded asscopy.
 	var/busy = FALSE
 
+/obj/machinery/photocopier/Initialize()
+	. = ..()
+	AddComponent(/datum/component/payment, 5, SSeconomy.get_dep_account(ACCOUNT_CIV), "Clinical")
```
Comment (bobbahbrown) [inline code/modules/paperwork/photocopier.dm:34]: defines

File: code/modules/paperwork/photocopier.dm [hunk 5]
```diff
@@ -56,6 +60,8 @@
 		return
 	if(href_list["copy"])
 		if(copy)
+			if(!SEND_SIGNAL(src, COMSIG_OBJ_ATTEMPT_CHARGE_PRINTER, usr))
```
Comment (bobbahbrown) [inline code/modules/paperwork/photocopier.dm:63]: did you intend on only having copies and not photocopies and doc copies cost money, as they are free below in the else ifs

File: code/datums/components/payment.dm [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  * What this is useful for:
+  * Basic forms of vending.
+  * Objects that can drain the owner's money linearly.
+  * What this is not useful for:
+  * Things where the seller may want to fluxuate the price of the object.
+  * Improving standardizing every form of payment handing, as some custom handling is specific to that object.
+  **/
+/datum/component/payment
+	///Standardized of operation.
+	var/cost = 10
```
Comment (bobbahbrown) [inline code/datums/components/payment.dm:13]: you can have dynamic pricing by just allowing to send a signal to override the charged amount, and hence make the component a lot more flexible

Comment (bobbahbrown) [reply code/datums/components/payment.dm:13]: in other words another signal that takes a 'cost' parameter which is used to attempt to charge the account with that cost

Comment (ArcaneMusic) [reply code/datums/components/payment.dm:13]: Oh yeah, that could very easily be done going forward, I was just hoping to put it off until more cases of the simple vendor can be worked out.

**Comment by TheMidnightRose** [comment]
Are you also fixing how photo copiers dont let you make more than one copy at a time?

**Comment by RaveRadbury** [comment]
If you're going to touch the photocopier can you _please_ fix the issue with printing multiple copies that popped up when `sleep` was refactored out of it?

**Comment by ArcaneMusic** [comment]
> Are you also fixing how photo copiers dont let you make more than one copy at a time?



> _Please_ fix the issue with printing multiple copies that popped up when `sleep` was refactored out of it.

Unbased and cringepilled, I'll see what I can do.

**Comment by ArcaneMusic** [comment]
Well the copier fix is done, I'm lazy and haven't done the defines so you'll see these below soon VVVV

File: code/datums/components/payment.dm [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	var/datum/bank_account/target
+
+/datum/component/payment/Initialize(_cost, _target, _style)
+	if(!_target)
+		target = SSeconomy.get_dep_account(ACCOUNT_CIV)
+	target = _target
+	cost = _cost
+	transaction_style = _style
+	RegisterSignal(parent, list(COMSIG_OBJ_ATTEMPT_CHARGE), .proc/attempt_charge)
+	RegisterSignal(parent, list(COMSIG_OBJ_ATTEMPT_CHARGE_CHANGE), .proc/change_cost)
```
Comment (Rohesie) [inline code/datums/components/payment.dm:27]:
```suggestion
	RegisterSignal(parent, COMSIG_OBJ_ATTEMPT_CHARGE, .proc/attempt_charge)
	RegisterSignal(parent, COMSIG_OBJ_ATTEMPT_CHARGE_CHANGE, .proc/change_cost)
```

File: code/datums/components/payment.dm [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+/datum/component/payment/proc/attempt_charge(datum/source, atom/movable/AM, var/extra_fees = 0)
+	var/mob/user = AM
```
Comment (Rohesie) [inline code/datums/components/payment.dm:30]:
```suggestion
/datum/component/payment/proc/attempt_charge(datum/source, atom/movable/target, var/extra_fees = 0)
	var/mob/user = target
```
If you have the guarantee that the user/target will be a mob, why are you not typecasting it earlier, before the signal is sent even?

File: code/datums/components/payment.dm [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	if(!cost) //In case a free variant of anything is made it'll skip charging anyone.
+		return
+	if(!istype(user))
+		return COMSIG_ITEM_CANCEL_CHARGE
+	var/obj/item/card/id/card = user.get_idcard(TRUE)
```
Comment (Rohesie) [inline code/datums/components/payment.dm:35]:
```suggestion
	if(!ismob(user))
		return COMSIG_ITEM_CANCEL_CHARGE
	var/mob/user = target
	var/obj/item/card/id/card = user.get_idcard(TRUE)
```

File: code/datums/components/payment.dm [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+				to_chat(user, "<span class='warning'>I'm so sorry... You don't seem to have enough money.</span>")
+			if(PAYMENT_ANGRY)
+				to_chat(user, "<span class='warning'>YOU MORON. YOU ABSOLUTE BAFOON. YOU INSUFFERABLE TOOL. YOU ARE POOR.</span>")
+			if(PAYMENT_CLINICAL)
+				to_chat(user, "<span class='warning'>ID Card lacks funds. Aborting.</span>")
+		return COMSIG_ITEM_CANCEL_CHARGE
+	target.transfer_money(card.registered_account, cost)
+	card.registered_account.bank_card_talk("[cost] credits deducted from your account.")
+	playsound(src, 'sound/effects/cashregister.ogg', 20, TRUE)
+	return
```
Comment (Rohesie) [inline code/datums/components/payment.dm:66]:
```suggestion

```

File: code/datums/components/payment.dm [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+/datum/component/payment/proc/change_cost(datum/source, var/new_cost)
+	if(!isnum(new_cost))
+		return
+	cost = new_cost
+	return
```
Comment (Rohesie) [inline code/datums/components/payment.dm:72]:
```suggestion
```
Return is implied.

File: code/datums/components/payment.dm [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+				to_chat(user, "<span class='warning'>ID Card lacks funds. Aborting.</span>")
+		return COMSIG_ITEM_CANCEL_CHARGE
+	target.transfer_money(card.registered_account, cost)
+	card.registered_account.bank_card_talk("[cost] credits deducted from your account.")
+	playsound(src, 'sound/effects/cashregister.ogg', 20, TRUE)
+	return
+
+/datum/component/payment/proc/change_cost(datum/source, var/new_cost)
+	if(!isnum(new_cost))
+		return
```
Comment (Rohesie) [inline code/datums/components/payment.dm:70]:
```suggestion
	if(!isnum(new_cost))
		CRASH("change_cost called with variable new_cost as not a number.")
```
If you add sanity checks, stuff you don't expect to normally fail, then add error messages so we don't have silent ones in the future when things go wrong.

File: code/game/machinery/medical_kiosk.dm [hunk 13]
```diff
@@ -29,35 +29,25 @@
...
 		use_power(20)
 		paying_customer = TRUE
 		say("Hello, esteemed medical staff!")
 		RefreshParts()
 		return
-	if(!account.has_money(active_price))
-		say("You do not possess the funds to purchase this.")  //No jobless crew, either.
+	var/bonus_fee = 0
+	if(pandemonium)
+		bonus_fee = rand(10,30)
```
Comment (Rohesie) [inline code/game/machinery/medical_kiosk.dm:46]:
```suggestion
	var/bonus_fee = pandemonium ? rand(10,30) : 0
```

File: code/__DEFINES/dcs/signals.dm [hunk 14]
```diff
@@ -482,13 +482,20 @@
 ///from [/obj/structure/closet/supplypod/proc/endlaunch]:
 #define COMSIG_SUPPLYPOD_LANDED "supplypodgoboom"
 
+// /obj signals for economy
+///called when the payment component tries to charge an account.
+#define COMSIG_OBJ_ATTEMPT_CHARGE "obj_attempt_simple_charge"
+	#define COMSIG_ITEM_CANCEL_CHARGE  (1<<0)
+///Called when a payment component changes value
+#define COMSIG_OBJ_ATTEMPT_CHARGE_CHANGE "obj_attempt_simple_charge_change"
```
Comment (Rohesie) [inline code/__DEFINES/dcs/signals.dm:490]: This doesn't seem to be ever used. What's that about?

Comment (ArcaneMusic) [reply code/__DEFINES/dcs/signals.dm:490]: This is more to set up a framework for replacing existing machinery that you _can charge_ people with, which includes machinery with variable prices.

**Pull request merged**